File: 515.在每个树行中找最大值.py
#
# @lc app=leetcode.cn id=515 lang=python
#
# [515] 在每个树行中找最大值
#
# 解法1(T81% S70%)：树的层序遍历，可以通过一个队列来维护，首先将(root, 0)初始化队列，其中0标记level，用来判断同层中最大的元素，之后while循环出队一个元素并判断是否为该层第一个元素或更大的元素，最后将left和right节点继续入队循环即可

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def largestValues(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root: return []
        
        queue = [(root, 0)]
        res = [root.val]

        while queue:
            node, level = queue.pop()
            if level > len(res)-1:
                res.append(node.val)
            if node.val > res[level]:
                res[level] = node.val

            if node.left:
                queue.append((node.left, level+1))
            if node.right:
                queue.append((node.right, level+1))
        return res
        
        
# @lc code=end

File: javascript/515.在每个树行中找最大值.js
/*
 * @lc app=leetcode.cn id=515 lang=javascript
 *
 * [515] 在每个树行中找最大值
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var largestValues = function(root) {
    if (!root) { return []; }
    var queue = [[root, 1]];
    var res = [];
    while (queue.length>0) {
        var data = queue.shift();
        var node=data[0], level=data[1];
        if (res.length < level) {
            res.push(node.val)
        } else if (res[res.length-1] < node.val) {
            res[res.length-1] = node.val
        }

        if (node.left) {
            queue.push([node.left, level+1]);
        }
        if (node.right) {
            queue.push([node.right, level+1]);
        }
    }
    return res;
};
// @lc code=end

